Bugfix pfs coalesce

charmster opened the pull request:
Fix bugs with the COALESCE operation for static-large objects where it dropped the metadata (headers) associated with the operation on the floor and failed to return the correct modification time.

Lack of headers broke multipart-upload in the S3API because the new object lacked the ACL headers so it could not be accessed via S3API.  It also dropped the S3API and SLO ETag values on the floor, which meant the object did not return the correct ETag later.

First commit: store selected metadata associated with the operation.  Mung the ETag headers os the request values are returned until the object is modified via SMB or NFS, at which point the ProxyFS generated ETag is returned.

Second commit: return the correct modification time for the COALESCE operaiton.

Commit:
Fix SLO to return correct "last-modified-time".

Change Coalesce() to add ctime (attr change time) to return values.

If the response to a Coalesce() request by middleware contains a ctime
value, it is use for "last-modified-time" in the REST response.  If not,
the mtime value is used.  This was added a few months ago as part of a
transition from mtime --> ctime for last-modifed-time.

CoalesceReply in RpcCoalesce() in jrpcfs/middleware.go includes a ctime
field, but it was not being set by anything.  Presumably its return
value was 0 and the middleware python code would then use the value of 0
for last-modified-time.

Plumb ctime as one of the values returned by Coalesce() through several
layers of the stack so it gets to the Python middleware code as it
should.

Note that we don't seem to have any end-to-end tests (swift-client to
Proxyfs) of SLO uploads (a PUT with "multipart-manifest=put") so we
don't have any tests that would detect this.

Commit:
Bugfix for PFS COALESCE/SLO/S3API multipart-upload for FileCatalyst

pfs_middleware and proxyfs ignored any metdata (headers) provided as
part of a COALESCE operation and did not store them, causing the
newly created object to lack S3 ACLs which meant it could not be
accessed or delted via S3API.

this fix stores any provided headers, other than `X-Static-Large-Object`
and headers that start with `X-Object-Sysmeta-Slo-`.

dropping the headers caused at least one bug--but may have avoided others.
this commit tries to fix the actual bug and the other theoretical bugs.

add code to mung and unmung the several ETag headers that could exist
and insure those ETags are dropped if the file changes via another API.

the bug that this fixes is for FileCatalyst uploads:  if the file being
uploaded already existed and it was created via the `s3api` multipart
upload then FileCatalyst would perform a GET (HEAD?) on the file, and
that would fail with a permission error because the s3Acl that should
have allowed to exceed was dropped on the floor.

Note: files uploaded via other APIs, such as file access or the Swift
object API will also lack an s3Acl and therefore cannot be deleted or
otherwise accessed via `s3api`.  this will likely cause other problems,
but it needs to addressed by a separate fix.

Fix the problem and add some tests to pfs_middleware to verify that
"good" headers are stored and "bad" headers are not stored.

Write bin/aws-multipart-test.py to be able to test the s3api
multipart-upload (SLO) against different accounts and to verify
that ETags and permissions (by implication) are stored correctly.
This test must be run manually and requires some setup.

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -438,30 +439,111 @@ def extract_container_metadata_from_headers(req):
+                del obj_metadata[LISTING_ETAG_OVERRIDE_HEADER]
         except ValueError:
+            del obj_metadata[LISTING_ETAG_OVERRIDE_HEADER]
             pass
```
nit: The pass is unnecessary now that we're doing something in the `except:` block.

I wonder if we could bring it closer to the original a bit, though; something like
```
    if LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
        val = obj_metadata.pop(LISTING_ETAG_OVERRIDE_HEADER)
        try:
            stored_num_writes, rest = val.split(':', 1)
            if int(stored_num_writes) == num_writes:
                obj_metadata[LISTING_ETAG_OVERRIDE_HEADER] = rest
        except ValueError:
            pass
```

and similar for the other headers.

charmster replied:
good point!  i'm not that good at python ... i'll fix that.

charmster replied:
sorry, i didn't address all of your comment!

anyway, if the conditional ETag header is no longer valid because somebody/something has written to the file, why do we want to keep it?  it just uses up disk space and network bandwidth.

is there a reason to keep `LISTING_ETAG_OVERRIDE_HEADER` once it is no longer germane?

tipabu replied:
Are we talking about keeping it persisted in the metadata stream in proxyfs? I'd assumed because we didn't want to have to go digging through the metadata to figure out what's still valid and what isn't on every write.

Or are we talking about in the request path? We definitely don't need to keep it around then; the `.pop()` will do just that:

```
>>> x={'foo': 'bar'}
>>> x
{'foo': 'bar'}
>>> x.pop('foo')
'bar'
>>> x
{}
```
```

charmster replied:
Ahhh.  yes, the .pop() is better.

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -438,30 +439,111 @@ def extract_container_metadata_from_headers(req):
+    return
+
+
+def unmung_etags_and_return_best(obj_metadata, account_name, inum, num_writes,
```
Should this should be broken up into two functions?
```
def unmung_etags(obj_metadata, num_writes):
```
which _only_ seeks to do the inverse of `mung_etags()`, and
```
def best_possible_etag(obj_metadata, account_name, inum,
                       is_dir=False, container_listing=False):
```
which expects `unmung_etags()` to have already been called and no longer needs `num_writes`...

charmster replied:
yes, it should probably be two functions.

i made this a single function to cut down the number of lines of code changed to make this change more palatable.

pretty much everywhere we call `best_possible_etag()` we have to call `unmung_etags()` first and there is no need to call `unmung_etags()` if you're not going to call `best_possible_etag()`.  so they are kind of "joined at the hip".

but i'm happy to change the code that looks like:
```            
          etag = unmung_etags_and_return_best(
                 obj_metadata, account_name,
                 container_entry["InodeNumber"],
                 container_entry["NumWrites"],
```
to:
```
            unmung_etags(obj_metadata)
            etag = best_possible_etag(
                 obj_metadata, account_name,
                 container_entry["InodeNumber"],
                 container_entry["NumWrites"],
```
i will do that.

charmster replied:
making this change turned out to be more complicated and, arguably, fixed a bug in POST and perhaps bugs where metadata was getting modified when it shouldn't have been (because i didn't make a copy of the dictionary when i probably should have).

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1471,9 +1554,11 @@ def put_object_as_directory(self, ctx):
-            "Etag": construct_etag(ctx.account_name, inode, num_writes),
             "Content-Type": DIRECTORY_CONTENT_TYPE,
             "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
+        resp_headers["Etag"] = unmung_etags_and_return_best(
```
Huh. Why isn't this _always_ `EMPTY_OBJECT_ETAG`? When we're making a directory, it's just a zero-byte object from Swift's perspective, right? But this was already kinda funky...

charmster replied:
the short answer is "i don't know."  the longer answer is "i don't know and i didn't really want to spend the time to figure it out."

from looking at the other code, i suspect that the PUT request that creates a directory could be coming from the `s3api` filter and it could supply its own value for the ETag, with the header name `S3API_ETAG_HEADER`.  there could also be a `LISTING_ETAG_OVERRIDE_HEADER`.

so, while a header named `ETag` (or `etag` or `Etag`) should have the value `EMPTY_OBJECT_ETAG`, there could also be a header named `S3API_ETAG_HEADER`, and if that header exists we should(?) return that ETag.

but, looking at the code for `unmung_etags_and_return_best()`, i can see that for a directory that function will only ever return `EMPTY_OBJECT_ETAG`.

i'm not sure if that's the correct behavior or not; i think i just copied what the earlier code seemed to do (see the above comment about "... and i don't really want to spend the time to figure it out.").

frankly, if the s3api had set a value for `LISTING_ETAG_OVERRIDE_HEADER` or `S3API_ETAG_HEADER` for the directory then i think `unmung_etags_and_return_best()` should return that ETag for the directory.

what do you think?

charmster replied:
i still think the correct behavior here is kind've murky.

_arguably_ if `S3API_ETAG_HEADER` was supplied with the request then that's what should be returned.  however, it is also true that the old `best_possible_etag()` code will return `EMPTY_OBJECT_ETAG` for all directories (or, more accurately, `EMPTY_OBJECT_ETAG` if the _optional_ `is_dir` is true; it default to false).

the original code seemed to have:
```
         resp_headers = {
-            "Etag": construct_etag(ctx.account_name, inode, num_writes),
             "Content-Type": DIRECTORY_CONTENT_TYPE,
             "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
```
which definitely seems wrong.

looking through the code, `get_object()` does not set the optional `is_dir` argument when calling `best_possible_etag()` and `get_object()` does not return the `IsDir` metadta..  however, the current behavior of proxyfsd is for a `get_object()` on a directory to fail with a `500` series error which seems quite wrong.

i have a fix for that (to fix problems that Burton is seeing) but it looks like that fix will need to percolate through the Python `get_object()` code.

anyway, in your opinion, should the ETag for a directory always be `EMPTY_OBJECT_ETAG` regardless of any headers set by the S3API layer?

charmster replied:
i changed the code again on the theory that `best_possible_etag()` might change in the future.  so now it looks like:
```
        # currently best_possible_etag() returns EMPTY_OBJECT_ETAG for                                                   
        # all directories, but that might change in the future.                                                          
        # unmung the obj_metadata so best_possible_etag() can use it                                                     
        # if its behavior changes (note that num_writes is forced to 0).                                                 
        unmung_etags(obj_metadata, 0)
        resp_headers = {
            "Content-Type": DIRECTORY_CONTENT_TYPE,
            "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
        resp_headers["ETag"] = best_possible_etag(
            obj_metadata, ctx.account_name, inode, 0, is_dir=True)
```

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1905,10 +1987,48 @@ def coalesce_object(self, ctx, auth_cb):
+        if req_etag is None:
+            req_etag = req.headers.get('ETAG', None)
+        if req_etag is None:
+            req_etag = req.headers.get('etag', None)
```
`req` is a `swob.Request`, right? All of these should be equivalent since `req.headers` is a [`HeaderEnvironProxy`](https://github.com/openstack/swift/blob/2.21.0/swift/common/swob.py#L236-L281) which translates them all into

    req.environ.get('HTTP_ETAG')

(Also, the `, None` should be unnecessary since we're using `.get()` where `None` is automatically the default.)

charmster replied:
i don't know anything about `swob.Request` and did not take the time to learn what it does (i was already having enough fun with the other inherited classes that seemed to affect how this code works).  i was fairly certain the code would work in all cases and not introduce any bugs.

from looking through the Swift code base, it seemed like there is a _diversity of opinion_ about capitalization:  `etag` could be lower-case (`etag`), upper-case (`ETAG`), title-case (`Etag`) or random case (`ETag`).  different filters actually seemed to convert the case of headers as the request passes through as they see fit!

so, if you're sure i can replace the existing code:
```
        # retrieve the ETag (or etag) value in the request (or None)                                                     
        req_etag = req.headers.get('Etag', None)
        if req_etag is None:
            req_etag = req.headers.get('ETag', None)
        if req_etag is None:
            req_etag = req.headers.get('ETAG', None)
        if req_etag is None:
            req_etag = req.headers.get('etag', None)
```
with
```
        # retrieve the ETag (or etag) value in the request (or None)                                                     
        req_etag = req.headers.get('Etag')
```
then i'm happy to make that change!

charmster replied:
i dug into this some more, looked at `swob.Request` and discovered that it eventually overloads dictionaries (maps?) with its own key munger `def header_to_environ_key(header_name)`, which does what you said.

there are some things that, having been seen, cannot be unseen.

but i changed the code to match your suggestion.

tipabu replied:
> there are some things that, having been seen, cannot be unseen.

lol! You should see what I'm having to do for py3... or maybe not, as the case may be.

FWIW, the opinions about ETag casing [get even stronger in s3api](https://github.com/openstack/swift/blob/2.21.0/swift/common/middleware/s3api/s3response.py#L34-L36). Some clients just don't understand that headers are supposed to be case-insensitive :frowning_face:

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1905,10 +1987,48 @@ def coalesce_object(self, ctx, auth_cb):
                         return denial_response
                 authed_containers.add(elem_container_path)
 
+        # treat the number of objects as the number of writes
```
It's not just that we're _treating it_ as such, right? It's that proxyfs **will** be using the number of coalesced files as the number of writes.

charmster replied:
well, yes and no!

formerly (before this patch) proxyfs tracked the number of writes for each object in the SLO and then added them together and returned that total to `parse_coalesce_object_response()`, which means it could have been different from the number of objects.

this patch also changes the proxyfs code to return (and store) the number of objects as the number of writes, instead of the total number of writes to create the component objects.

since the result of the COALESCE is a new object, i think we can assign any number of writes we want (including "1") and the ETag code will still work correctly.  so perhaps i should have changed this code, and corresonding proxyfs code, to make this `1`.

so i changed the way num_writes is calculated by proxyfs `Coalesce()`  to try and make the whole operation more transactional.  

if i don't know `num_writes` until `parse_coalesce_object_response()` returns, then i can't mung the metadata until then and i would have to perform a second transaction to set the correct metadata after the coalesce transaction, which opens the door (or window) to it never getting done (if the node crashes between updates).

so i took a sort of "middle of the road" approach my using the number of objects as the number of writes instead of the actual number of writes or simply using 1.

in my opinion, the comment is fine ... what do you think i should change the comment to say?

charmster replied:
i updated the comment to:
```
        # proxyfs treats the number of objects as the number of writes                                                   
```

tipabu commented on inode/file.go:
```diff
@@ -1187,16 +1193,21 @@ func (vS *volumeStruct) Coalesce(destInodeNumber InodeNumber, elements []*Coales
+	// attach new middleware headers (why are we making a copy?)
+	inodeStreamBuf := make([]byte, len(metaData))
+	copy(inodeStreamBuf, metaData)
+	destInode.StreamMap[metaDataName] = inodeStreamBuf
```
This will all be fine when `metaData` is `nil`, right? Just thinking about rolling upgrades, when you may have an old proxy talking to an upgraded proxyfs...

charmster replied:
Good catch!

curiously, while an uninitialized  slice is nil, taking the length of it still returns 0.

>  The zero value of a slice is nil. The len and cap functions will both return 0 for a nil slice. 

from https://blog.golang.org/go-slices-usage-and-internals

i tried it out in the Go playground, including using the gob encoder/decoder to see how it did with fields that are not sent and it seems fine.

https://play.golang.org/p/Qgtd_Ai6cD1

tipabu commented on pfs_middleware/tests/test_pfs_middleware.py:
```diff
@@ -4072,6 +4136,26 @@ def mock_RpcCoalesce(coalese_req):
+        self.assertNotIn('X-Object-Sysmeta-Slo-Etag', headers)
+        self.assertNotIn('X-Object-Sysmeta-Slo-Size', headers)
+
+        print("test_success(): headers: '%s'\n" % (headers))
```
Might want to take out the print.

charmster replied:
yes, the `print()` is now gone.

tipabu commented on pfs_middleware/tests/test_pfs_middleware.py:
```diff
@@ -4438,33 +4522,33 @@ class TestBestPossibleEtag(unittest.TestCase):
     def test_modified(self):
         self.assertEqual(
-            mware.best_possible_etag(
+            mware.unmung_etags_and_return_best(
```
Since this now has two responsibilities, we should probably keep a reference to the dict and check that `self.HEADER` is no longer present after the call.

charmster replied:
i had to rewrite that code again and found i had to copy() dictionaries to get it to work right.  i think that addresses this comment ...

tipabu reviewed (commented):
As a heads-up, I mostly looked at the pfs_middleware code, not so much the Go code. And I haven't actually functionally tested to verify all the plumbing.

I think I like the general approach, but
- I might prefer better separation of concerns with `unmung_etags_and_return_best()`,
- there's some redundancy when we go to get the COALESCE ETag, and
- we should probably have better testing of the unmunging of valid metadata and removal of invalid metadata.

We might also want to switch `aws-multipart-test.py` to use `boto3` directly, but that seems beyond the scope of what ought to happen before a merge.

Commit:
Incorporate changes suggested in code review.

Break the functionality of unmung_etags_and_return_best() into
unmung_etags() and return_best_possible_etag() and propagate the changes
through the code.

Along the way, i noticed there were probably some bugs with when and
how we "unmunged" the etags which are fixed.  In addition, post_object()
did not seem to handle the case where the POST added one of the "special"
etags that was not previously present (if it was previously present the
POST would not be able to change it).

put_object_as_directory() now returns an ETag of EMPTY_OBJECT_ETAG which
means the corresponding unit test had to be changed; i think the orginal
behavior was incorrect.

Replace class TestBestPossibleEtag tests with class TestEtagHandling,
which has separate tests for mung_etags(), unmung_etags(),
best_possible_etag().  The tests now run through all three ETag headers.

charmster commented:
@tipabu can i get you take a look at my last set of changes and see if all of your concerns are resolved?

the two most controversial behaviors are probably this code in POST (post_object()) which allows certain ETAG values to be set if they were not set previously (which, i believe, is the previous behavior):

https://github.com/swiftstack/ProxyFS/blob/9174a329a59d70f76d1df0e486ebc74cc91e53e8/pfs_middleware/pfs_middleware/middleware.py#L1724-L1743

and these comments about the way ETAGs are contstructed which, again, i think matches existing behavior:
https://github.com/swiftstack/ProxyFS/blob/9174a329a59d70f76d1df0e486ebc74cc91e53e8/pfs_middleware/tests/test_pfs_middleware.py#L4569-L4581

but which leads to the need for this code:
https://github.com/swiftstack/ProxyFS/blob/9174a329a59d70f76d1df0e486ebc74cc91e53e8/pfs_middleware/pfs_middleware/middleware.py#L1736-L1739

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1462,18 +1543,29 @@ def put_object_as_directory(self, ctx):
+
+        # mung the passed etags, if any (NumWrites for a directory is
+        # always zero)
+        mung_etags(obj_metadata, request_etag, 0)
```
:+1: good call

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1635,8 +1721,22 @@ def post_object(self, ctx):
+        # supply a value S3API_ETAG_HEADER or LISTING_ETAG_OVERRIDE_HEADER
+        # (or ORIGINAL_MD5_HEADER?).  While merge_object_metadata() will
+        # not let these overwrite an existing header, they can be added to
+        # an object that doesn't have such a header.
```
Hmmm.... we shouldn't be adding any new sysmeta on a POST.... but it looks like this is pre-existing behavior. We should maybe fix it at some point, but doesn't have to be part of this.

charmster replied:
OK, i wasn't sure of the correct semantics.

i will leave this as existing behavior.

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1635,8 +1721,22 @@ def post_object(self, ctx):
+        #
+        # Note that if an ETag header is stale (because NumWrites has
+        # changed) then unmung_etags() will strip it and it will not be
+        # put back by mung_etags().
```
Yeah, that should be fine. Since the underlying data changed, we have no use for it any more.

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -1829,6 +1932,7 @@ def head_object(self, ctx):
             rpc.parse_head_response(head_response)
 
         headers = swob.HeaderKeyDict(deserialize_metadata(raw_md))
+        unmung_etags(headers, num_writes)
```
This works equally-well with both `dict`s and `HeaderKeyDict`s, right?

charmster replied:
i have no idea!  i will need to investigate ...

Commit:
Further fixes for Etags and random case headers.

A client of pfs_middleware will no longer be able to add
"X-Object-Sysmeta-*" or "X-Static-Large-Object" headers via a POST
operation on an object.

A client of pfs_middleware will no longer be able to slip in a
"X-Object-Sysmeta-ProxyFS-Initial-MD5" header or a
"X-Object-Sysmeta-Proxyfs-Initial-Md5" header to a PUT operation.

Fix a bug where coalesce_object() would fail to return a caller
supplied ETag value.

tipabu commented on pfs_middleware/pfs_middleware/middleware.py:
```diff
@@ -402,7 +417,13 @@ def extract_object_metadata_from_headers(headers):
+
+            # do not let a client pass in ORIGINAL_MD5_HEADER
+            if (header != ORIGINAL_MD5_HEADER and
+                header != ORIGINAL_MD5_HEADER.title()):
```
I think `tox -e lint` will be happier with something like
```
if header != ORIGINAL_MD5_HEADER and \
        header != ORIGINAL_MD5_HEADER.title():
    meta_headers[header] = value
```
though maybe it'd be more pythonic as
```
if header not in (ORIGINAL_MD5_HEADER,
                  ORIGINAL_MD5_HEADER.title()):
    meta_headers[header] = value
```
? The reduced repetition might even let it all fit on one line.
